### DISCUSSION TITLE
pg_bm25 from clause compilation

### DISCUSSION THREAD

**Discussion by DeoLeung** [sqlalchemy/sqlalchemy#10841] [Usage Questions]
Hi, for postgresql plugin `pg_bm25`
* index definition https://docs.paradedb.com/latest/indexing/bm25
```sql
paradedb.create_bm25(
  index_name => '<index_name>',
  table_name => '<table_name>',
  key_field => '<key_field>'
  text_fields => '<text_fields>',
  numeric_fields => '<numeric_fields>',
  boolean_fields => '<boolean_fields>',
  json_fields => '<json_fields>',
);

```
* searching https://docs.paradedb.com/latest/search/bm25
```sql
SELECT *
FROM <index_name>.search('<query>');
```

For core usage, it's there any idea about how to define the index, and use it as `from` clause?

**Comment by CaselIT** [comment]
Hi,

Without doing anything custom these both seem function calls, so you van use `func` to render them.

To use a function in a from there is `table_valued`

**Reply by DeoLeung** [reply]
thank you, I make it work for searching , use something like
```python
from sqlalchemy import func
from sqlalchemy import select

print(select(func.my_table_search_idx.search('abc').table_valued('id', 'name')))
```

how about the index creation function, especially the `=>` part

**Reply by CaselIT** [reply]
that's just how kwargs are rendered in postgresql. Does it work if you use python kwargs in the function?

**Reply by DeoLeung** [reply]
it seems not working
```python
print(
    func.paradedb.create_bm25(index_name='abc_idx').compile(
        dialect=postgresql.dialect(),
        compile_kwargs={
            "literal_binds": True,
        },
    ))
```
only prints the function name, I custom compile the `Function`
```python
@compiles(Function)
def compile_kw(element, compiler, **kw):
  print(element.name, element.__dict__)
```
```bash
create_bm25 {'packagenames': ('paradedb',), 'name': 'create_bm25', '_bind': None, 'type': NullType(), '_has_args': False, 'clause_expr': <sqlalchemy.sql.elements.Grouping object at 0x1049190d0>, 'clauses': <sqlalchemy.sql.elements.ClauseList object at 0x104918e90>, '_memoized_keys': frozenset({'clauses'})}
```
seems not storing the kwargs?

**Reply by CaselIT** [reply]
seems not. You will have to call that function positionally.
This should be equivalent to the above
```sql
paradedb.create_bm25(
  '<index_name>',
  '<table_name>',
  '<key_field>'
  '<text_fields>',
  '<numeric_fields>',
  '<boolean_fields>',
  '<json_fields>',
);
```

@zzzeek do you think allowing a function to be called using kwargs would make sense?

**Reply by zzzeek** [reply]
I would propose a PG-specific construct for "name based function", or something that varies `func.` upfront.   i dont want to complicate the default `FunctionGenerator`

**Reply by CaselIT** [reply]
I'm not sure if this is a pg only thing or it's more general to other db. 

Looking a bit it's also an oracle thing, since 11g: https://logicalread.com/oracle-11g-named-parameters-in-sql-function-calls-mc02/

Same syntax even. (but I think pg accepts different ways, may be wrong here)

**Reply by zzzeek** [reply]
something like `func.named_parameters.foobar()` or something

**Reply by CaselIT** [reply]
yes, that was my thoughts too if we were to add something to the general api.

I'm not sure if it's worth it though

**Reply by DeoLeung** [reply]
just found a workaround which use `text` to do it a bit, before it's added to general api.
```python
from sqlalchemy import func
from sqlalchemy import select
from sqlalchemy import text

print(select(
    func.my_table_search_idx.search('abc').table_valued('id', 'name')))

print(
    func.paradedb.create_bm25(text("index_name => 'my_table_search_idx'"),
                              text("table_name => 'my_table'")))
```

**Reply by DeoLeung** [reply]
in case someone needs it before kwargs added to sqlalchemy, I created a small [helper function](https://gist.github.com/DeoLeung/ebd3f14c96cfc95ad0a025caaf114b3d) for complete usage
